### PR TITLE
T19028

### PR DESCRIPTION
**Pull request by tchx84**
https://phabricator.endlessm.com/T19028

### CODE REVIEW

File: tools/incubator.js [hunk 1]
```diff
@@ -32,6 +33,19 @@ const Incubator = new Lang.Class({
         return file.get_path();
     },
 
+    _download_resource: function (uri) {
+        print('Downloading', uri);
+
+        let session = new Soup.Session();
+        let message = Soup.Message.new('GET', uri);
+        let input_stream = session.send(message, null);
```
Comment (tchx84) [inline tools/incubator.js]: 4myself: need to do something when this fails...

File: lib/basin-util.c [hunk 2]
```diff
@@ -95,6 +112,26 @@ basin_override_resources (const gchar *content)
     }
   xmlXPathFreeObject(objects);
 
+  /* override videos links */
+  objects = xmlXPathEvalExpression ("//a[@data-libingester-asset-id]", context);
+  if (!xmlXPathNodeSetIsEmpty (objects->nodesetval))
+    {
+      htmlDocPtr img_node = htmlReadDoc (THUMBNAIL_SVG, "", NULL, HTML_PARSE_RECOVER | HTML_PARSE_NOERROR);
```
Comment (tchx84) [inline lib/basin-util.c]: 4myself: I could do this in a similar but simpler way.

**Comment by tchx84** [comment]
@ptomato I added 2 comments for myself, things I forgot to add. But please go throught it.. I want your opinion on things like not actually extracting thumbnail and using a placeholder like I do now.

File: lib/basin-util.c [hunk 3]
```diff
@@ -95,6 +112,26 @@ basin_override_resources (const gchar *content)
...
+      xmlNodeSetPtr nodeset = objects->nodesetval;
+      for (int i = 0; i < nodeset->nodeNr; i++)
+        {
+          ekn_id = xmlGetProp (nodeset->nodeTab[i], "data-libingester-asset-id");
+          prop = g_strdup_printf ("ekn:///%s", ekn_id);
+          xmlSetProp (nodeset->nodeTab[i], "href", prop);
+          xmlFree (prop);
```
Comment (ptomato) [inline lib/basin-util.c]: Should be `g_free()` — probably both call `free()` in practice, but always free API-allocated memory with the free function advised by that API

Comment (ptomato) [reply lib/basin-util.c]: (And this is a pre-existing defect, but strictly speaking the elements of the list returned by `basin_find_resources()` should be freed with `xmlFree()`, but GJS has no way to know that ... so you should probably copy them with `g_strdup()` and free the original XML characters)

File: tools/incubator.js [hunk 4]
```diff
@@ -32,6 +33,21 @@ const Incubator = new Lang.Class({
         return file.get_path();
     },
 
+    _download_file: function (uri) {
+        let session = new Soup.Session();
+        let message = Soup.Message.new('GET', uri);
+        let input_stream = session.send(message, null);
```
Comment (ptomato) [inline tools/incubator.js]: I think you can accomplish the same thing without Soup. You can just create a GFile for `uri`, `read()` it, and splice it to the temp file's output stream.

File: examples/desc.json [hunk 5]
```diff
@@ -9,16 +9,15 @@
...
+    },
     {
         "contentType": "image/jpg",
         "license": "CC-BY 3.0",
         "tags": ["EknMediaObject"],
         "height": 531,
         "width": 540,
-        "source": "image.jpg"
+        "source": "image.jpg",
+        "notIndexed": true
```
Comment (ptomato) [inline examples/desc.json]: `"indexed": false`? :smile:

**Review by ptomato — CHANGES_REQUESTED** [review]
Hmm, the placeholder is a bit bad because there's nothing that the user can do even if they want to extract a thumbnail themselves. Maybe you could make the placeholder slightly more user friendly than a black SVG (e.g., a black SVG with a play icon) and provide an option in the manifest for the user to add a thumbnail if they are prepared to edit the manifest before using basin on it.

**Comment by tchx84** [comment]
@ptomato we already provide the "play icon" svg, it comes from the article format CSS. It looks good IMHO.

screenshot from 2017-09-07 11-25-13

File: lib/basin-util.c [hunk 6]
```diff
@@ -32,6 +34,21 @@ basin_find_resources (const gchar *content)
...
+
+  /* from videos */
+  objects = xmlXPathEvalExpression ("//a[@data-libingester-asset-id]", context);
+  if (!xmlXPathNodeSetIsEmpty (objects->nodesetval))
+    {
+      xmlNodeSetPtr nodeset = objects->nodesetval;
+      for (int i = 0; i < nodeset->nodeNr; i++)
+        {
+          xmlChar *ekn_id = xmlGetProp (nodeset->nodeTab[i], "data-libingester-asset-id");
+          list = g_list_append (list, g_strdup (ekn_id));
```
Comment (ptomato) [inline lib/basin-util.c]: Also a pre-existing problem, but I didn't spot it last time: `g_list_append()` is inefficient when adding multiple items, use `g_list_prepend()` and `g_list_reverse()` afterwards if the order matters.

File: tools/incubator.js [hunk 7]
```diff
@@ -85,6 +98,25 @@ const Incubator = new Lang.Class({
         return basin_metadata;
     },
 
+    _import_video: function (asset) {
+        let basin_metadata = new Map();
+
+        let file = this._download_file(asset['uri']);
+        let info = file.query_info('*', Gio.FileQueryInfoFlags.NONE, null);
+
+        basin_metadata['source'] = file.get_path();
```
Comment (ptomato) [inline tools/incubator.js:115]: Just realized another pre-existing defect that I didn't spot last time: you're setting these keys as properties on the `Map` object, and not as keys in the map's hashtable. To access the hashtable, you have to use `get()`, `set()`, and `has()`.

File: tools/incubator.js [hunk 8]
```diff
@@ -32,6 +32,19 @@ const Incubator = new Lang.Class({
         return file.get_path();
     },
 
+    _download_file: function (uri) {
+        let source_file = Gio.File.new_for_uri(uri);
```
Comment (ptomato) [inline tools/incubator.js]: Hmm, maybe this was a bad suggestion on my part. I now get an "Operation not supported" exception thrown when I do this.

**Review by ptomato — CHANGES_REQUESTED** [review]

File: tools/incubator.js (hunk 8 above)
Comment (tchx84) [reply tools/incubator.js]: Is this on endlessOS? I had that issue too, but it got fixed after installing glib-networking and gvfs on jhbuild.

File: tools/incubator.js (hunk 7 above)
Comment (tchx84) [reply tools/incubator.js:115]: ouch, I feel very silly. Can I fix this all-in-once in a posterior commit?

File: tools/incubator.js (hunk 8 above)
Comment (ptomato) [reply tools/incubator.js]: I was using a flapjack runtime.

File: tools/incubator.js (hunk 7 above)
Comment (ptomato) [reply tools/incubator.js:115]: :+1:

File: lib/basin-util.c (hunk 6 above)
Comment (tchx84) [reply lib/basin-util.c]: Thanks!

File: tools/incubator.js (hunk 8 above)
Comment (tchx84) [reply tools/incubator.js]: I can just mimic the same behavior, with GFile.is_native and Soup, no big deal.

**Review by ptomato — APPROVED** [review]
OK. I didn't have time to test it in a flapjack sandbox this time, since my local flapjack is in a non-working state at the moment, but I won't block the PR on that.

**Pull request merged**